README.md
# helm-open-github.el [![melpa badge][melpa-badge]][melpa-link] [![melpa stable badge][melpa-stable-badge]][melpa-stable-link]
helm-open-github.el is utilities for opening github url.
This is inspired by URL below.

- http://shibayu36.hatenablog.com/entry/2013/01/18/211428


## Screenshot

![open-github-from-commit](image/helm-open-github-from-commit.png)


## Installation

You can install `helm-open-github` from [MELPA](http://melpa.org/) with package.el.

```
 M-x package-install helm-open-github
```


## Dependency

* Emacs 24.3 or higher
* [helm](https://github.com/emacs-helm/helm)
* [gh](https://github.com/sigma/gh.el)


## Basic Usage

### `helm-open-github-from-commit`

Open commit page from **Commit ID**

### `helm-open-github-from-file`

Open file page from **File Name**

### `helm-open-github-from-issues`

Open issue page from **Issue ID**

### `helm-open-github-from-pull-requests`

Open pull request page from **Pull Request ID**


## Customize

### `helm-open-github-commit-limit`

Issue number shown by `helm-open-github-from-commit`.(Default: `100`)

### `helm-open-github-requires-pattern`

Minimal length to search. If this value is non-nil, delayed search is enabled(Default: `nil`).
This parameter must be set before loading `helm-open-github.el`.


## Sample Configuration

```lisp
(global-set-key (kbd "C-c o f") 'helm-open-github-from-file)
(global-set-key (kbd "C-c o c") 'helm-open-github-from-commit)
(global-set-key (kbd "C-c o i") 'helm-open-github-from-issues)
(global-set-key (kbd "C-c o p") 'helm-open-github-from-pull-requests)
```

[melpa-link]: https://melpa.org/#/helm-open-github
[melpa-stable-link]: https://stable.melpa.org/#/helm-open-github
[melpa-badge]: https://melpa.org/packages/helm-open-github-badge.svg
[melpa-stable-badge]: https://stable.melpa.org/packages/helm-open-github-badge.svg


Changes
Revision history for helm-open-github.el

Revision 0.15 2016/12/03 syohex
   - Require Emacs 24.4 or higher versions
   - Use newer helm interfaces

Revision 0.14 2015/10/10 syohex
   - Fix too many candidates issue(#12 Thanks thierryvolpiatto)
   - Update minimum helm version
   - Fix buffer name by helm way(#15 Thanks xuchunyang)

Revision 0.13 2015/03/03 syohex
   - Fix wrong 'gh' version #9(Reported by nlamirault)

Revision 0.12 2015/01/17 syohex
   - Improve issue command(Thanks thierryvolpiatto)
     We can see all issues now.

Revision 0.11 2014/07/02 syohex
   - Re-factoring

Revision 0.10 2014/07/02 syohex
   - Re-factoring

Revision 0.09 2014/06/21 syohex
   - Fix for cl-lib
   - Use Cask now

Revision 0.08 2014/03/08 syohex
   - Enable lexical binding
   - switch cl-lib

Revision 0.07 2013/11/14 syohex
   - Fix missing requirements(Thanks tdksk)

Revision 0.06 2013/11/14 syohex
   - Implement pull request interface

Revision 0.05 2013/10/26 syohex
   - Fix repository regexp
   - Support multiple candidates

Revision 0.04 2013/05/01 syohex
   - Implement persistent action to helm-open-github-from-commit

Revision 0.03 2013/04/30 syohex
   - Fix name attribute of helm-open-github--from-file-source
     (Thanks shibayu36)

Revision 0.02 2013/04/30 syohex
   - Implement show commit detail

Revision 0.01 2013/04/27 syohex
   - Initial release


open-github.el
;;; open-github.el --- Utilities of Opening Github Page

;; Copyright (C) 2013 by Syohei YOSHIDA

;; Author: Syohei YOSHIDA <syohex@gmail.com>
;; URL: https://github.com/syohex/emacs-open-github
;; Version: 0.01
;; Package-Requires: ((helm "1.0") (gh "1.0"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; Open github URL utilities. This package is inspired by URL below.
;;   - http://shibayu36.hatenablog.com/entry/2013/01/18/211428

;;; Code:

(eval-when-compile
  (require 'cl))

(require 'helm)
(require 'gh-issues)

(defgroup open-github nil
  "Utilities of opeg "
  :prefix "open-github-"
  :group 'http)

(defcustom open-github-commit-limit 100
  "Limit of commit id collected"
  :type 'integer
  :group 'open-github)

(defcustom open-github-issues-api
  (gh-issues-api "api" :sync t :cache nil :num-retries 1)
  "Github API instance. This is-a `gh-issues'"
  :group 'open-github)

(defun open-github--collect-commit-id ()
  (let ((cmd (format "git --no-pager log -n %d --pretty=oneline --abbrev-commit"
                     open-github-commit-limit)))
    (with-current-buffer (helm-candidate-buffer 'global)
      (let ((ret (call-process-shell-command cmd nil t)))
        (unless (zerop ret)
          (error "Failed: git log(retval=%d)" ret))))))

(defun open-github--command-one-line (cmd)
  (with-temp-buffer
    (let ((ret (call-process-shell-command cmd nil t)))
      (when (zerop ret)
        (goto-char (point-min))
        (buffer-substring-no-properties
         (line-beginning-position) (line-end-position))))))

(defun open-github--full-commit-id (abbrev-id)
  (let ((cmd (concat "git rev-parse " abbrev-id)))
    (or (open-github--command-one-line cmd)
        (error "Failed: %s" cmd))))

(defun open-github--root-directory ()
  (let ((root (open-github--command-one-line "git rev-parse --show-toplevel")))
    (if (not root)
        (error "Error: here is not Git repository")
      (file-name-as-directory root))))

(defun open-github--host ()
  (or (open-github--command-one-line "git config --get hub.host")
      "github.com"))

(defun open-github--remote-url ()
  (let ((cmd "git config --get remote.origin.url"))
    (or (open-github--command-one-line cmd)
        (error "Failed: %s" cmd))))

(defun open-github--extract-user-host (remote-url)
  (if (string-match "[:/]\\([^/]+\\)/\\([^/]+?\\)\\.git\\'" remote-url)
      (values (match-string 1 remote-url) (match-string 2 remote-url))
    (error "Failed: match %s" remote-url)))

(defun open-github--commit-url (host remote-url commit-id)
  (multiple-value-bind (user repo) (open-github--extract-user-host remote-url)
    (format "https://%s/%s/%s/commit/%s"
            host user repo commit-id)))

(defun open-github--from-commit-action-common (commit-id)
  (let* ((host (open-github--host))
         (remote-url (open-github--remote-url)))
    (browse-url
     (open-github--commit-url host remote-url commit-id))))

(defun open-github--from-commit-action (line)
  (let* ((commit-line (split-string line " "))
         (commit-id (open-github--full-commit-id (car commit-line))))
    (open-github--from-commit-action-common commit-id)))

(defun open-github--from-commit-direct-input-action (unused)
  (let ((commit-id (read-string "Input Commit ID: ")))
    (open-github--from-commit-action-common
     (open-github--full-commit-id commit-id))))

(defvar open-github--from-commit-source
  '((name . "Open Github From Commit")
    (init . open-github--collect-commit-id)
    (candidates-in-buffer)
    (action . open-github--from-commit-action)))

(defvar open-github--from-commit-direct-input-source
  '((name . "Open Github From Commit Direct Input")
    (candidates . ("Input Commit ID"))
    (action . open-github--from-commit-direct-input-action)))

;;;###autoload
(defun open-github-from-commit ()
  (interactive)
  (helm :sources '(open-github--from-commit-source
                   open-github--from-commit-direct-input-source)
        :buffer "*open github*"))

(defun open-github--collect-files ()
  (let ((root (open-github--root-directory)))
    (with-current-buffer (helm-candidate-buffer 'global)
      (let* ((default-directory root)
             (cmd "git ls-files")
             (ret (call-process-shell-command cmd nil t)))
        (unless (zerop ret)
          (error "Failed: %s(%s)" cmd default-directory))))))

(defun open-github--branch ()
  (let* ((cmd "git symbolic-ref HEAD")
         (branch (open-github--command-one-line cmd)))
    (if (not branch)
        (error "Failed: %s" cmd)
      (replace-regexp-in-string "\\`refs/heads/" "" branch))))

(defun open-github--file-url (host remote-url branch file marker)
  (multiple-value-bind (user repo) (open-github--extract-user-host remote-url)
    (format "https://%s/%s/%s/blob/%s/%s%s"
            host user repo branch file marker)))

(defun open-github--highlight-marker (start end)
  (cond ((and start end)
         (format "#L%s..L%s" start end))
        (start
         (format "#L%s" start))
        (t "")))

(defun open-github--from-file-action (file &optional start end)
  (let ((host (open-github--host))
        (remote-url (open-github--remote-url))
        (branch (open-github--branch))
        (marker (open-github--highlight-marker start end)))
    (browse-url
     (open-github--file-url host remote-url branch file marker))))

(defun open-github--from-file-highlight-region-action (file)
  (let ((start-line (read-number "Start Line: "))
        (end-line (read-number "End Line: ")))
    (open-github--from-file-action file start-line end-line)))

(defun open-github--from-file-highlight-line-action (file)
  (let ((start-line (read-number "Start Line: ")))
    (open-github--from-file-action file start-line)))

(defvar open-github--from-file-source
  '((name . "Open Github From Commit")
    (init . open-github--collect-files)
    (candidates-in-buffer)
    (action . (("Open File" .
                (lambda (file) (open-github--from-file-action file)))
               ("Open File and Highlight Line"
                . open-github--from-file-highlight-line-action)
               ("Open File and Highlight Region"
                . open-github--from-file-highlight-region-action)))))

(defun open-github--from-file-direct (file start end)
  (let* ((root (open-github--root-directory))
         (repo-path (file-relative-name file root))
         (start-line (line-number-at-pos start))
         (end-line (line-number-at-pos end)))
    (open-github--from-file-action repo-path start-line end-line)))

;;;###autoload
(defun open-github-from-file ()
  (interactive)
  (if mark-active
      (open-github--from-file-direct (buffer-file-name) (region-beginning) (region-end))
    (helm :sources '(open-github--from-file-source)
          :buffer "*open github*")))

(defun open-github--collect-issues ()
  (let ((host (open-github--host))
        (remote-url (open-github--remote-url)))
    (multiple-value-bind (user repo) (open-github--extract-user-host remote-url)
      (let ((issues (gh-issues-issue-list open-github-issues-api user repo)))
        (if (null issues)
            (error "This repository has no issues!!")
          (sort (oref issues data)
                (lambda (a b) (< (oref a number) (oref b number)))))))))

(defun open-github--convert-issue-api-url (url)
  (replace-regexp-in-string
   "api\\." ""
   (replace-regexp-in-string "/repos" "" url)))

(defun open-github--from-issues-real-to-display (issue)
  (with-slots (number title state) issue
    (format "#%-4d [%s] %s" number state title)))

(defvar open-github--from-issues-source
  '((name . "Open Github From Issues")
    (candidates . open-github--collect-issues)
    (volatile)
    (real-to-display . open-github--from-issues-real-to-display)
    (action . (lambda (issue)
                (browse-url
                 (open-github--convert-issue-api-url (oref issue url)))))))

(defun open-github--construct-issue-url (host remote-url issue-id)
  (multiple-value-bind (user repo) (open-github--extract-user-host remote-url)
    (format "https://%s/%s/%s/issues/%s"
            host user repo issue-id)))

(defun open-github--from-issues-direct (host)
  (let ((remote-url (open-github--remote-url))
        (issue-id (read-number "Issue ID: ")))
    (browse-url
     (open-github--construct-issue-url host remote-url issue-id))))

;;;###autoload
(defun open-github-from-issues ()
  (interactive)
  (let ((host (open-github--host)))
    (if (not (string= host "github.com"))
        (open-github--from-issues-direct host)
      (helm :sources '(open-github--from-issues-source)
            :buffer  "*open github*"))))

(provide 'open-github)

;;; open-github.el ends here


helm-open-github.el
;;; helm-open-github.el --- Utilities of Opening Github Page -*- lexical-binding: t; -*-

;; Copyright (C) 2016 by Syohei YOSHIDA

;; Author: Syohei YOSHIDA <syohex@gmail.com>
;; URL: https://github.com/syohex/emacs-helm-open-github
;; Version: 0.15
;; Package-Requires: ((emacs "24.4") (helm-core "1.7.7") (gh "0.8.2"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; Open github URL utilities. This package is inspired by URL below.
;;   - http://shibayu36.hatenablog.com/entry/2013/01/18/211428

;;; Code:

(require 'cl-lib)

(require 'helm)
(require 'gh-issues)
(require 'gh-pulls)

(defgroup helm-open-github nil
  "Utilities of opeg "
  :prefix "helm-open-github-"
  :group 'http)

(defcustom helm-open-github-commit-limit 100
  "Limit of commit id collected"
  :type 'integer)

(defcustom helm-open-github-issues-api
  (gh-issues-api "api" :sync t :cache nil :num-retries 1)
  "Github API instance. This is-a `gh-issues'"
  :type 'gh-issues-api)

(defcustom helm-open-github-pulls-api
  (gh-pulls-api "api" :sync t :cache nil :num-retries 1)
  "Github API instance. This is-a `gh-pulls'"
  :type 'gh-pulls-api)

(defcustom helm-open-github-closed-issue-since 19
  "Only issues updated this number of days ago are returned."
  :type 'integer)

(defcustom helm-open-github-closed-issue-sort-direction "asc"
  "Direction of the sort for closed issues.
Either \"asc\" or \"desc\"."
  :type '(radio :tag "Preferred direction of the sort"
          (const :tag "Ascendent" "asc")
          (const :tag "Descendent" "desc")))

(defcustom helm-open-github-requires-pattern nil
  "Minimal length to search. As fetching data is an expensive
operation with potentially many results, higher number is
recomended for bigger projects or slower connections.
If this value is non-nil, delayed search is disabled."
  :type '(choice (integer :tag "Minimal length")
                 (boolean :tag "Disable delayed search" nil)))

(defun helm-open-github--collect-commit-id ()
  (with-current-buffer (helm-candidate-buffer 'global)
    (let ((ret (call-process "git" nil t nil
                             "--no-pager" "log"
                             "-n" (number-to-string helm-open-github-commit-limit)
                             "--pretty=oneline" "--abbrev-commit")))
      (unless (zerop ret)
        (error "Failed: get commit ID")))))

(defun helm-open-github--command-one-line (cmd args)
  (with-temp-buffer
    (when (zerop (apply 'call-process cmd nil t nil args))
      (goto-char (point-min))
      (buffer-substring-no-properties
       (line-beginning-position) (line-end-position)))))

(defun helm-open-github--full-commit-id (abbrev-id)
  (or (helm-open-github--command-one-line "git" `("rev-parse" ,abbrev-id))
      (error "Failed: 'git rev-parse %s'" abbrev-id)))

(defun helm-open-github--root-directory ()
  (let ((root (helm-open-github--command-one-line "git" '("rev-parse" "--show-toplevel"))))
    (if (not root)
        (error "Error: here is not Git repository")
      (file-name-as-directory root))))

(defun helm-open-github--host ()
  (or (helm-open-github--command-one-line "git" '("config" "--get" "hub.host"))
      "github.com"))

(defun helm-open-github--remote-url ()
  (or (helm-open-github--command-one-line "git" '("config" "--get" "remote.origin.url"))
      (error "Failed: Can't get remote.origin URL")))

(defun helm-open-github--extract-user-host (remote-url)
  (if (string-match "[:/]\\([^/]+\\)/\\([^/]+?\\)\\(?:\\.git\\)?\\'" remote-url)
      (cl-values (match-string 1 remote-url) (match-string 2 remote-url))
    (error "Failed: match %s" remote-url)))

(defun helm-open-github--commit-url (host remote-url commit-id)
  (cl-multiple-value-bind (user repo) (helm-open-github--extract-user-host remote-url)
    (format "https://%s/%s/%s/commit/%s"
            host user repo commit-id)))

(defun helm-open-github--from-commit-open-url-common (commit-id)
  (let* ((host (helm-open-github--host))
         (remote-url (helm-open-github--remote-url)))
    (browse-url
     (helm-open-github--commit-url host remote-url commit-id))))

(defun helm-open-github--full-commit-id-from-candidate (line)
  (let ((commit-line (split-string line " ")))
    (helm-open-github--full-commit-id (car commit-line))))

(defun helm-open-github--from-commit-id-persistent-action (line)
  (let* ((commit-id (helm-open-github--full-commit-id-from-candidate line))
         (str (shell-command-to-string
               (format "git show --stat --oneline %s" commit-id))))
    (with-help-window (help-buffer)
      (princ str))))

(defun helm-open-github--from-commit-open-url (_candidate)
  (dolist (commit-line (helm-marked-candidates))
    (let ((commit-id (helm-open-github--full-commit-id-from-candidate commit-line)))
      (helm-open-github--from-commit-open-url-common commit-id))))

(defun helm-open-github--from-commit-open-url-with-input (_candidate)
  (let ((commit-id (read-string "Input Commit ID: ")))
    (helm-open-github--from-commit-open-url-common
     (helm-open-github--full-commit-id commit-id))))

(defun helm-open-github--show-commit-id-common (commit-id)
  (with-current-buffer (get-buffer-create "*open-github-issues*")
    (unless (call-process "git" nil t nil "show" "--stat" "-p" commit-id)
      (error "Error: 'git show --stat -p %s'" commit-id))
    (goto-char (point-min))
    (pop-to-buffer (current-buffer))))

(defun helm-open-github--show-commit-id (line)
  (let* ((commit-line (split-string line " "))
         (commit-id (helm-open-github--full-commit-id (car commit-line))))
    (helm-open-github--show-commit-id-common commit-id)))

(defun helm-open-github--show-commit-id-with-input (_candidate)
  (let ((commit-id (read-string "Input Commit ID: ")))
    (helm-open-github--show-commit-id-common
     (helm-open-github--full-commit-id commit-id))))

(defvar helm-open-github--from-commit-source
  (helm-build-in-buffer-source "Open Github From Commit"
    :init #'helm-open-github--collect-commit-id
    :persistent-action #'helm-open-github--from-commit-id-persistent-action
    :action (helm-make-actions
             "Open Commit Page" #'helm-open-github--from-commit-open-url
             "Show Detail" #'helm-open-github--show-commit-id)))

(defvar helm-open-github--from-commit-direct-input-source
  (helm-build-sync-source "Open Github From Commit Direct Input"
   :candidates '("Input Commit ID")
   :action (helm-make-actions
            "Open Commit Page" #'helm-open-github--from-commit-open-url-with-input
            "Show Detail" #'helm-open-github--show-commit-id-with-input)))

;;;###autoload
(defun helm-open-github-from-commit ()
  (interactive)
  (helm :sources '(helm-open-github--from-commit-source
                   helm-open-github--from-commit-direct-input-source)
        :buffer "*helm open github*"))

(defun helm-open-github--collect-files ()
  (let ((root (helm-open-github--root-directory)))
    (with-current-buffer (helm-candidate-buffer 'global)
      (let ((default-directory root))
        (unless (zerop (call-process "git" nil t nil "ls-files"))
          (error "Failed: 'git ls-files' at %s" default-directory))))))

(defun helm-open-github--branch ()
  (let ((branch (helm-open-github--command-one-line "git" '("symbolic-ref" "HEAD"))))
    (if (not branch)
        (error "Failed: 'git symbolic-ref HEAD'")
      (replace-regexp-in-string "\\`refs/heads/" "" branch))))

(defun helm-open-github--file-url (host remote-url branch file marker)
  (cl-multiple-value-bind (user repo) (helm-open-github--extract-user-host remote-url)
    (format "https://%s/%s/%s/blob/%s/%s%s"
            host user repo branch file marker)))

(defun helm-open-github--highlight-marker (start end)
  (cond ((and start end)
         (format "#L%s..L%s" start end))
        (start
         (format "#L%s" start))
        (t "")))

(defun helm-open-github--from-file-action (file &optional start end)
  (let ((host (helm-open-github--host))
        (remote-url (helm-open-github--remote-url))
        (branch (helm-open-github--branch))
        (marker (helm-open-github--highlight-marker start end)))
    (browse-url
     (helm-open-github--file-url host remote-url branch file marker))))

(defun helm-open-github--from-file-highlight-region-action (file)
  (let ((start-line (read-number "Start Line: "))
        (end-line (read-number "End Line: ")))
    (helm-open-github--from-file-action file start-line end-line)))

(defun helm-open-github--from-file-highlight-line-action (file)
  (let ((start-line (read-number "Start Line: ")))
    (helm-open-github--from-file-action file start-line)))

(defvar helm-open-github--from-file-source
  (helm-build-in-buffer-source "Open Github From File"
    :init #'helm-open-github--collect-files
    :action (helm-make-actions
             "Open File" (lambda (_cand)
                           (dolist (file (helm-marked-candidates))
                             (helm-open-github--from-file-action file)))
             "Open File and Highlight Line" #'helm-open-github--from-file-highlight-line-action
             "Open File and Highlight Region" #'helm-open-github--from-file-highlight-region-action)))

(defun helm-open-github--from-file-direct (file start end)
  (let* ((root (helm-open-github--root-directory))
         (repo-path (file-relative-name file root))
         (start-line (line-number-at-pos start))
         (end-line (line-number-at-pos end)))
    (helm-open-github--from-file-action repo-path start-line end-line)))

;;;###autoload
(defun helm-open-github-from-file ()
  (interactive)
  (if mark-active
      (helm-open-github--from-file-direct (buffer-file-name) (region-beginning) (region-end))
    (helm :sources '(helm-open-github--from-file-source)
          :buffer "*helm open github*")))

(defun helm-open-github--collect-issues ()
  (let ((remote-url (helm-open-github--remote-url)))
    (cl-multiple-value-bind (user repo) (helm-open-github--extract-user-host remote-url)
      (let ((issues (gh-issues-issue-list helm-open-github-issues-api user repo)))
        (if (null issues)
            (error "This repository has no issues!!")
          (sort (oref issues data)
                (lambda (a b) (> (oref a number) (oref b number)))))))))

(defmethod gh-issues-issue-list-closed ((api gh-issues-api) user repo)
  (gh-api-authenticated-request
   api (gh-object-list-reader (oref api issue-cls)) "GET"
   (format "/repos/%s/%s/issues" user repo)
   nil `(("state" . "closed")
         ("since" . ,(format-time-string
                      "%y-%m-%dT%H:%M:%SZ"
                      (time-subtract
                       (current-time)
                       (seconds-to-time
                        (* (* (* helm-open-github-closed-issue-since 24)
                              60) 60)))))
         ("direction" . ,helm-open-github-closed-issue-sort-direction))))

(defun helm-open-github--collect-closed-issues ()
  (let ((remote-url (helm-open-github--remote-url)))
    (cl-multiple-value-bind (user repo) (helm-open-github--extract-user-host remote-url)
      (let ((issues (gh-issues-issue-list-closed helm-open-github-issues-api user repo)))
        (if (null issues)
            (error "This repository has no issues!!")
            (oref issues data))))))

(defun helm-open-github--convert-issue-api-url (url)
  (replace-regexp-in-string
   "api\\." ""
   (replace-regexp-in-string "/repos" "" url)))

(defun helm-open-github--from-issues-format-candidate (issue)
  (with-slots (number title state) issue
    (propertize (format "#%-4d [%s] %s" number state title)
                'helm-realvalue issue)))

(defun helm-open-github--open-issue-url (_candidate)
  (dolist (issue (helm-marked-candidates))
    (browse-url (oref issue html-url)
                (helm-open-github--convert-issue-api-url (oref issue url)))))

(defvar helm-open-github--issues-cache (make-hash-table :test 'equal))
(defvar helm-open-github--from-issues-source
  (helm-build-in-buffer-source "Open Github From Open Issues"
    :init (lambda ()
            (let* ((key (helm-open-github--remote-url))
                   (issues (gethash key helm-open-github--issues-cache)))
              (unless issues
                (setq issues
                      (puthash key (helm-open-github--collect-issues)
                               helm-open-github--issues-cache)))
              (helm-init-candidates-in-buffer 'global
                (cl-loop for c in issues
                         collect (helm-open-github--from-issues-format-candidate c)))))
    :delayed (not (null helm-open-github-requires-pattern))
    :requires-pattern helm-open-github-requires-pattern
    :get-line 'buffer-substring
    :action (helm-make-actions
             "Open issue page with browser" #'helm-open-github--open-issue-url)))

(defvar helm-open-github--closed-issues-cache (make-hash-table :test 'equal))
(defvar helm-open-github--from-closed-issues-source
  (helm-build-in-buffer-source "Open Github From closed Issues"
    :init (lambda ()
            (let* ((key (helm-open-github--remote-url))
                   (issues (gethash key helm-open-github--closed-issues-cache)))
              (unless issues
                (setq issues
                      (puthash key (helm-open-github--collect-closed-issues)
                               helm-open-github--closed-issues-cache)))
              (helm-init-candidates-in-buffer 'global
                (cl-loop for c in issues
                         collect (helm-open-github--from-issues-format-candidate c)))))
    :delayed (not (null helm-open-github-requires-pattern))
    :requires-pattern helm-open-github-requires-pattern
    :get-line 'buffer-substring
    :action (helm-make-actions
             "Open issue page with browser" #'helm-open-github--open-issue-url)))

(defun helm-open-github--construct-issue-url (host remote-url issue-id)
  (cl-multiple-value-bind (user repo) (helm-open-github--extract-user-host remote-url)
    (format "https://%s/%s/%s/issues/%s"
            host user repo issue-id)))

(defun helm-open-github--from-issues-direct (host)
  (let ((remote-url (helm-open-github--remote-url))
        (issue-id (read-number "Issue ID: ")))
    (browse-url
     (helm-open-github--construct-issue-url host remote-url issue-id))))

;;;###autoload
(defun helm-open-github-from-issues (arg)
  (interactive "P")
  (let ((host (helm-open-github--host))
        (url (helm-open-github--remote-url)))
    (when arg
      (remhash url helm-open-github--closed-issues-cache)
      (remhash url helm-open-github--issues-cache))
    (if (not (string= host "github.com"))
        (helm-open-github--from-issues-direct host)
      (helm :sources '(helm-open-github--from-issues-source
                       helm-open-github--from-closed-issues-source)
            :buffer  "*helm open github*"))))

(defvar helm-open-github--pull-requests nil)
(defun helm-open-github--collect-pullreqs ()
  (let ((remote-url (helm-open-github--remote-url)))
    (cl-multiple-value-bind (user repo) (helm-open-github--extract-user-host remote-url)
      (let ((pullreqs (gh-pulls-list helm-open-github-pulls-api user repo)))
        (if (null pullreqs)
            (error "This repository has no pull requests!!")
          (setq helm-open-github--pull-requests
                (sort (oref pullreqs data)
                      (lambda (a b) (< (oref a number) (oref b number))))))))))

(defun helm-open-github--pulls-view-common (url)
  (with-current-buffer (get-buffer-create "*open-github-diff*")
    (view-mode -1)
    (erase-buffer)
    (unless (zerop (call-process "curl" nil t nil "-s" url))
      (error "Can't download %s" url))
    (goto-char (point-min))
    (diff-mode)
    (view-mode +1)
    (pop-to-buffer (current-buffer))))

(defun helm-open-github--pulls-view-diff (candidate)
  (helm-open-github--pulls-view-common (oref candidate diff-url)))

(defun helm-open-github--pulls-view-patch (candidate)
  (helm-open-github--pulls-view-common (oref candidate patch-url)))

(defvar helm-open-github--from-pulls-source
  (helm-build-sync-source "Open Github From Issues"
    :init #'helm-open-github--collect-pullreqs
    :candidates 'helm-open-github--pull-requests
    :volatile t
    :delayed (not (null helm-open-github-requires-pattern))
    :requires-pattern helm-open-github-requires-pattern
    :real-to-display 'helm-open-github--from-issues-format-candidate
    :action (helm-make-actions
             "Open issue page with browser" #'helm-open-github--open-issue-url
             "View Diff" #'helm-open-github--pulls-view-diff
             "View Patch" #'helm-open-github--pulls-view-patch)))

;;;###autoload
(defun helm-open-github-from-pull-requests ()
  (interactive)
  (let ((host (helm-open-github--host)))
    (if (not (string= host "github.com"))
        (helm-open-github--from-issues-direct host)
      (helm :sources '(helm-open-github--from-pulls-source)
            :buffer  "*helm open github*"))))

(provide 'helm-open-github)

;;; helm-open-github.el ends here
